Best Practices and Anti-Patterns Updates (#1471)
onflow/cadence: docs/design-patterns.mdx

This is a selection of software design patterns developed by core Flow developers
while writing Cadence code for deployment to Flow Mainnet.

Many of these design patters apply to most other programming languages, but some are specific to Cadence.

[Design patterns](https://en.wikipedia.org/wiki/Software_design_pattern) are building blocks for software development.
They may provide a solution to a problem that you encounter when writing smart contracts in Cadence.
If they do not clearly fit, these patterns may not be the right solution for a given situation or problem.
They are not meant to be rules to be followed strictly, especially where a better solution presents itself.

# General

These are general patterns to follow when writing smart contracts.

## Use Named Value Fields for Constants Instead of Hard-Coding

### Problem

Your contracts, resources, and scripts all have to refer to the same value.
A number, a string, a storage path, etc.
Entering these values manually in transactions and scripts is a potential source of error.

> See: [https://en.wikipedia.org/wiki/Magic*number*(programming)](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)

Reviewer: ```suggestion
See [Wikipedia's page on magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming))
```

### Solution

> Add a public (`pub`) field, e.g. a `Path` , to the contract responsible for the value,

Reviewer: ```suggestion
Add a public (`pub`), constant (`let`) field, e.g. a `Path`, to the contract responsible for the value,
```

    init() {
        // assign and access the field here and in transactions
        self.TestStoragePath = /storage/testStorage
        self.account.save(<-create Test(), to: self.TestStoragePath)
    }
}

```

[Example Code](https://github.com/onflow/flow-core-contracts/blob/master/contracts/LockedTokens.cdc#L583)

## Script-Accessible Public Field/Function

Data availability is important in a blockchain environment.
It is useful to publicize information about your smart contract and the assets it controls
so other smart contracts and apps can easily query it.

### Problem

Your contract, resource or struct has a field or resource that will need to be read and used on or off-chain, often in bulk.

### Solution

Make sure that the field can be accessed from a script (using a `PublicAccount`)
rather than requiring a transaction (using an `AuthAccount`).
This saves the time and soon the expense required by having to read a property using a transaction.
Making the field or function `pub` and exposing it via a `/public/` capability will allow this.

Be careful not to expose any data or functionality that should be kept private when doing so.

Example:

```cadence

> // Field is private, so it cannot be read by the public
> // Bad
> access(self) let totalSupply: UFix64
> 
> // Field is public, so it can be read and used by anyone
> // Good
> pub let totalSupply: UFix64

Reviewer: ```suggestion
// BAD: Field is private, so it cannot be read by the public
access(self) let totalSupply: UFix64

// GOOD: Field is public, so it can be read and used by anyone
pub let totalSupply: UFix64
```

```

## Script-Accessible Report

### Problem

Your contract has a resource that you wish to access fields of.
Resources are often stored in private places and are hard to access.
Additionally, scripts cannot return resources to the external context,
so a construct must be used to hold the data.

### Solution

Declare a struct to hold the data that you wish to return from the script.

> Write a function that fills out the fields of this struct with the data

Reviewer: I think returning a reference should work and will have less boilerplate. It depends on how much data is needed off-chain

    pub resource BResource {
        pub var c: UInt64
        pub var d: string

        // Generate a struct with the same fields
        // to return when a script wants to see the fields of the resource
        // without having to return the actual resource
        pub fun generateReport() BReportStruct {
            return BReportStruct(c: c, d: d)
        }

        init(c: UInt64, d: string) {
            self.c = c
            self.d = d
        }
    }

    // Define a struct with the same fields as the resource
    pub struct BReportStruct {
        pub var c: UInt64
        pub var d: string

        init(c: UInt64, d: string) {
            self.c = c
            self.d = d
        }

    }
}
...
import A from 0xA

// Return the struct with a script
pub fun main(): A.BReport {
    let b: AContract.BResource // Borrow the resource
    let report = b.generateReport()
    return b
}
```

## Init Singleton

### Problem

An admin resource must be created and delivered to a specified account.
There should not be a function to do this, as that would allow anyone to create an admin resource.

### Solution

Create any one-off resources in the contract's `init()` function

> and deliver them to an address or AuthAccount specified as an argument.

Reviewer 2: ```suggestion
and deliver them to an address or `AuthAccount` specified as an argument.
```

It is usually safe to include post-conditions in transactions to verify the intended outcome.

### Example

This could be used when purchasing an NFT to verify that the NFT was deposited in your account's collection.

```cadence
// Psuedo-code

transaction {

    pub let buyerCollectionRef: &NonFungibleToken.Collection

    prepare(acct: AuthAccount) {

        // Get tokens to buy and a collection to deposit the bought NFT to
        let temporaryVault <- vaultRef.withdraw(amount: 10.0)
        let self.buyerCollectionRef = acct.borrow(from: /storage/Collection)

        // purchase, supplying the buyers collection reference
        saleRef.purchase(tokenID: 1, recipient: self.buyerCollectionRef, buyTokens: <-temporaryVault)

    }
    post {
        // verify that the buyer now owns the NFT
        self.buyerCollectionRef.idExists(1) == true: "Bought NFT ID was not deposited into the buyers collection"
    }
}
```

## Avoid Excessive Load and Save Storage Operations (Prefer in-place Mutations)

### Problem

When modifying data in account storage, `load()` and `save()` are costly operations.
This can quickly cause your transaction to reach the gas limit or slow down the network.

> ### Solution

Reviewer: Maybe also describe how this is not limited to direct storage API usage, but also e.g. contract objects and their fields (which are implicitly stored in storage, i.e. read from/written to), or nested resourced. 

For example, a collection contains a dictionary of NFTs. There is no need to move the whole dictionary out of the field, update the dictionary on the stack (e.g. adding or removing an NFT), and then move the whole dictionary back to the field, it can be updated in-place. The same goes for a more complex data structure like the sharded collection: It can be updated in-place by taking a reference to the nested object.

This reference can be assigned to or can be used to access fields or call methods on stored objects.

### Example

```cadence
// Loads and stores a resource to use it
// BAD
transaction {

    prepare(acct: AuthAccount) {

        // Removes the vault from storage, a costly operation
        let vault <- acct.load<@ExampleToken.Vault>(from: /storage/exampleToken)

        // Withdraws tokens
        let burnVault <- vault.withdraw(amount: 10)

        destroy burnVault

        // Saves the used vault back to storage, another costly operation
        acct.save(to: /storage/exampleToken)

    }
}

// Uses borrow instead to avoid costly operations
// GOOD
transaction {

    prepare(acct: AuthAccount) {

        // Borrows a reference to the stored vault, much less costly operation
        let vault <- acct.borrow<&ExampleToken.Vault>(from: /storage/exampleToken)

        let burnVault <- vault.withdraw(amount: 10)

        destroy burnVault

        // No `save` required because we only used a reference
    }
}
```

# Capabilities

## Capability Receiver

### Problem

An account must be given a [capability](/cadence/language/capability-based-access-control/)

> to a resource or contract in another account but a single transaction with both accounts authorizing it
> cannot be easily produced. This prevents a single transaction from fetching the capability

Reviewer: It might not be clear to the reader why the transaction must be signed by both parties. Maybe explain:

```suggestion
to a resource or contract in another account. To create, i.e. link the capability, the transaction must be signed by a key which has access to the target account.
To transfer / deliver the capability to the other account, the transaction also needs write access to that one.
It is not as easy to produce a single transaction which is authorized by two accounts as it is to produce a typical transaction which is authorized by one account. 
This prevents a single transaction from fetching the capability
```

This ensures that nobody else can create instances of it and call the receiver function
on B's receiver capability instead of A. This means that A is probably an admin account.

The second is that the field on the receiver resource that stores the desired Capability
should be `access(contract)` and only accessed by code within the contract that needs to.
This ensures that B cannot copy and share the Capability with anyone else.

### Example Code

See:

[LockedTokens.cdc](https://github.com/onflow/flow-core-contracts/blob/bfb115869bd9f815cde1fe64ab6d91ca95c0938e/contracts/LockedTokens.cdc#L527-L546)

[custody_setup_account_creator.cdc](https://github.com/onflow/flow-core-contracts/blob/79941fe65b634800065a440ae5243744b2ca8a2f/transactions/lockedTokens/admin/custody_setup_account_creator.cdc)

[admin_deposit_account_creator.cdc](https://github.com/onflow/flow-core-contracts/blob/79941fe65b634800065a440ae5243744b2ca8a2f/transactions/lockedTokens/admin/admin_deposit_account_creator.cdc)

## Capability Revocation

### Problem

A capability provided by one account to a second account must able to be revoked
by the first account without the co-operation of the second.

> ### Solution

Reviewer: Maybe add a callout and link to the [Capability Controller FLIP](https://github.com/onflow/flow/pull/798), which aims to improve this
